component/encryptor/encryptor.go

```diff
@@ -0,0 +1,382 @@
+package encryptor
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/config"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+	"github.com/Azure/azure-storage-fuse/v2/internal/handlemap"
+)
+
+type Encryptor struct {
+	internal.BaseComponent
+	handle           *os.File
+	blockSize        uint64
+	mountPointCipher string
+	encryptionKey    []byte
+}
+
+type LastChunkMeta struct {
+	sync.Mutex
+	farthestBlockSeen int64 // keep track of the farthest block that has been written to the file
+	paddingLength     int64 // keep track of the padding length of the last block
+}
+
+var lastchunkMeta = &LastChunkMeta{
```

Comment:
>lastchunkMeta [](http://example.com/codeflow?start=4&length=13)

At what granularity is lastChunkMeta supposed to operate at? Does this make it like a static/global across all blobs?

Reply:
This is at file level. Yes i agree it makes more sense to add this in Encryptor struct.

Reply:
was a global variable actually a bug also? trying to understand how it worked for 2 files in parallel.

Reply:
locking was to handle parallel writes in single file (write for 2nd block could come before 1st block). Wasn't sure if parallel file upload is a valid scenario for us so that is not yet handled.